### DISCUSSION TITLE
Loading meshes

### DISCUSSION THREAD

**Discussion by bhaveshshrimali** [adtzlr/felupe#622] [General]
Hi @adtzlr ,

Is it possible to load an external mesh via meshio (I do see that [here](https://github.com/adtzlr/felupe/blob/f9eaf63458f5bcc532164b81f461ca6a7e7b0ceb/src/felupe/mesh/_read.py#L29)) but not sure, e.g.

```python
from meshio import xdmf
import felupe as fem

meshio_mesh = xdmf.read("trialMesh.xdmf")
mesh = fem.Mesh(
    points=meshio_mesh.points,
    cells=meshio_mesh.get_cells_type("tetra"),
    cell_type=meshio_mesh.cells[0].type,
)

```

However this complains the cell volumes being less than zero for some of the cells, I tried flipping as suggested, namely,

```python
region_u = fem.RegionQuadraticTetra(mesh=mesh)
mesh = mesh.flip(np.any(region_u.dV < 0, axis=0))
region_u = fem.RegionQuadraticTetra(mesh=mesh)
region_p = fem.RegionTetra(
    mesh=fem.mesh.dual(mesh, disconnect=False),
    quadrature=region_u.quadrature,
    grad=False,
)
```
but this seems to just invert the ones having positive volume to negative. Would you know how to properly load an external mesh, or could point me to an example

For a test mesh I used:
```python

from skfem import MeshTet
mesh = MeshTet().refined(2)
mesh.save("trialMesh.xdmf")
```

**Comment by adtzlr** [comment]
Hi, I created an issue #624. This is fixed in #625.

```python
from skfem import MeshTet

mesh_skfem = MeshTet().refined(2)
mesh_skfem.save("trialMesh.xdmf")

from meshio import xdmf
import felupe as fem
import numpy as np

meshio_mesh = xdmf.read("trialMesh.xdmf")
mesh = fem.Mesh(
    points=meshio_mesh.points,
    cells=meshio_mesh.get_cells_type("tetra"),
    cell_type=meshio_mesh.cells[0].type,
)

# ensure positive cell volumes (before adding midpoints on edges!)
region = fem.RegionTetra(mesh)
mesh = mesh.flip(np.any(region.dV < 0, axis=0))
region.reload(mesh)
assert np.all(region.dV > 0)

region_u = fem.RegionQuadraticTetra(mesh.add_midpoints_edges())
region_p = fem.RegionTetra(
    mesh=fem.mesh.dual(mesh, disconnect=False),
    quadrature=region_u.quadrature,
    grad=False,
)
```

P.S.: It is also possible to use [`container = mesh.read(filename)`](https://felupe.readthedocs.io/en/latest/felupe/mesh.html#felupe.mesh.read) and select the desired [mesh](https://felupe.readthedocs.io/en/latest/felupe/mesh.html#felupe.Mesh) from the list of meshes of the [mesh container](https://felupe.readthedocs.io/en/latest/felupe/mesh.html#felupe.MeshContainer).

```python
>>> import felupe as fem

>>> container = fem.mesh.read(filename="trialMesh.xdmf")
>>> container.meshes
[<felupe Mesh object>
   Number of points: 115
   Number of cells:
     tetra: 320]

>>> container.meshes[0].imshow()
```

![image](https://github.com/adtzlr/felupe/assets/5793153/d7fca422-24c8-4c81-b6dc-eb744ecedb09)

**Reply by bhaveshshrimali** [reply]
wonderful, thanks a lot

**Comment by adtzlr** [comment]
I just pushed [v7.18.0](https://github.com/adtzlr/felupe/releases/tag/v7.18.0) to PyPI with the fix included.
